runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -419,6 +724,82 @@ impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API>
 impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API> module::BlockHandler
     for Module<Accounts, Consensus>
 {
+    fn end_block<C: Context>(ctx: &mut C) {
+        // Only do work in case the epoch has changed since the last processed block.
+        if !<C::Runtime as Runtime>::Core::has_epoch_changed(ctx) {
+            return;
+        }
+
+        let mut reclaims: lru::LruCache<(EpochTime, Address), (u128, u128)> =
+            lru::LruCache::new(NonZeroUsize::new(128).unwrap());
+
+        let own_address = Address::from_runtime_id(ctx.runtime_id());
+        let denomination = Consensus::consensus_denomination(ctx).unwrap();
+        let qd = state::get_queued_undelegations(ctx, ctx.epoch()).unwrap();
+        for ud in qd {
+            let udi = state::take_undelegation(ctx, &ud).unwrap();
+
+            // Determine total amount the runtime got during the reclaim operation.
+            let (total_amount, total_shares) =
+                if let Some(totals) = reclaims.get(&(ud.epoch, ud.from)) {
+                    *totals
+                } else {
+                    // Fetch consensus height corresponding to the given epoch transition. This
+                    // query may be expensive in case the epoch is far back, but the node is
+                    // guaranteed to have it as it was the state after the last normal round
+                    // (otherwise we would have already processed this epoch).
+                    let height = Consensus::height_for_epoch(ctx, ud.epoch)
+                        .expect("failed to determine height for epoch");
+
+                    // Find the relevant reclaim escrow event.
+                    let totals = ctx
+                        .history()
+                        .consensus_events_at(height, EventKind::Staking)
+                        .expect("failed to fetch historic events")
+                        .iter()
+                        .find_map(|ev| match ev {
```

Comment:
Maybe add a comment here that there will always be exactly one matching reclaim escrow event here, because debonding delegations get merged in consensus (if there are multiple reclaims for the same accounts on the same epoch).